### ISSUE TITLE
Simplified component signatures

### ISSUE THREAD

**Issue by natefaubion** [slamdata/purescript-halogen#288]
Right now, state is totally opaque to the outside world, yet we still have to thread around the composed type. Additionally, parent eval functions have to thread around the child state even though nothing can be done with it. If we want to commit to this route, with a small change we can eliminate a lot of the type boilerplate. The following is essentially the same thing we do now, but because we move the initial state into the component definition, we no longer have to deal with threading the type around, so we can focus on the only interface we actually expose: the query algebra.

``` purescript
type Render s f g =
  { state :: s
  , hooks :: Array (Hook f g)
  , tree  :: Tree f Unit
  }

-- By moving `initialState` to the component spec we can existentially hide it.
-- Components should be parameterized if they need to take a "custom" initial state.
type ComponentImpl s f g =
  { render :: s -> Render s f g
  , eval :: Natural f (ComponentDSL s f g)
  , initialState :: s
  , initializer :: Maybe (f Unit)
  , finalizers :: s -> Array (Finalized g)
  }

foreign import data Component :: (* -> *) -> (* -> *) -> *

mkComponent :: forall s f g. ComponentImpl s f g -> Component f g

runComponent :: forall f g r. (forall s. ComponentImpl s f g -> r) -> Component f g -> r

type InstalledImpl s f g =
  { component :: ComponentImpl s f g
  , state :: s
  , memo :: Maybe (Unit -> Tree f Unit)
  }

foreign import data Installed :: (* -> *) -> (* -> *) -> *

mkInstalled :: forall s f g. InstalledImpl s f g -> Installed f g

runInstalled :: forall f g r. (forall s. InstalledImpl s f g -> r) -> Installed f g -> r

newtype ParentState s f' p g = ParentState
  { state :: s
  , children :: Map p (Installed f' g)
  }

type ComponentHTML f = HTML Void (f Unit)

type ComponentDSL s f g = Free (HalogenF s f g)

type ParentQuery f f' p = Coproduct f (ChildF p f')

type ParentHTML f f' p g = HTML (Slot f' p g) (f Unit)

type ParentDSL s f f' p g = Free (HalogenFP ParentEventSource s f (QueryF s f' p g))

type QueryF s f' p g = Free (HalogenF (ParentState s f' p g) (ChildF p f') g)

data Slot f p g = Slot p (Component f g)

data ChildF p f i = ChildF p (f i)

---

data Query

someComponent :: Component Query (Aff ...)
someComponent = component { render, eval, initialState } where

  initialState :: Int

  render :: Int -> ComponentHTML Query

  eval :: Natural Query (ComponentDSL Int Query (Aff ...))


data PQuery

anotherComponent :: Component (ParentQuery PQuery Query Int) (Aff ...)
anotherComponent = parentComponent { render, eval, initialState } where

  initialState :: String

  render :: String -> ParentHTML PQuery Query Int (Aff ...)

  eval :: Natural PQuery (ParentDSL String PQuery Query Int (Aff ...))
```

So basically, the user doesn't ever have to know or write `ParentState`. Parent eval's have the expected type: parent state, parent query, child query, slot, base functor. Is this a direction we want to move in, or do we want to figure out how we can open up state more? If we did this #221 would become impossible.

**Comment by garyb** [comment]
This looks pretty great to me! We'd have to update the `slot` and `slot'` functions to just thunk the component construction and omit the `initialState`, but that would be nicer too.

The main difficulty I see is how `transform`ing a component would work with some of the types being existentially hidden now.

**Comment by jdegoes** [comment]
I agree that hiding state is a natural stepping stone on the current path, and will simplify a lot of types; so long as a component always has its initial state, then in theory, the state can be existentially hidden. I do believe that some of the embedding code might have to rely on coercion, simply because the type system isn't strong enough to make the kinds of guarantees we need; but I also think the code could be (manually) proved to be safe.

My only objection is that it's incompatible with #221. 

The main goals of #221 are as follows:

1. Allow parents to easily access children state, which eliminates a lot of unnecessary querying (you achieve this by passing in a child selector as well as the parent's nest);
2. Eliminate the messiness and deep nesting of coproduct hierarchies, and replace it with a much flatter, user-defined query algebra, whose individual terms are projected via prisms.

I think (2) can still be achieved with hiding the state, but definitely not (1). 

But, that said, current Halogen probably doesn't want you to have direct access to the state (across component boundaries), because it's used (or can be used) for caching the view. This sometimes leads to much more complex code, with parents querying to get the child state when they could simply look at it instead, but of course achieves stronger encapsulation and may benefit from caching.

So I think it could go either way. It comes down to, What does current Halogen really want to be?

**Comment by garyb** [comment]
In SlamData I don't think the situation (1) is trying to help with occurs very often... I can't think of any component I worked on that ended up having to request state from children actually. `peek` probably takes care of a lot of cases where (1) might otherwise occur.

**Comment by natefaubion** [comment]
> The main difficulty I see is how transforming a component would work with some of the types being existentially hidden now.

Is this actually an issue? The reason you have to transform state is to do all the `Either` (un)wrapping so all sub-components can be stored together. If this state is hidden, you don't actually have to transform it, I don't think.

But some other thoughts...

You can recover the current formulation of components with
```purescript
newtype ComponentS s f g = ComponentS (s -> Component f g)
```

I think if you took it further, in a rather hand-wavy sort of way, you could also add an output state and recover the `Profunctor` stuff from earlier versions. Except now I think we have the plumbing to make it work. I'd imagine you could write a combinator to then loop the state back to the parent. This would then be equivalent to having separate view state and model state. Recovering a closed off `Component f g`  (only view state) would then be a matter of having a `Unit` model state that could be ignored.

**Comment by garyb** [comment]
> Is this actually an issue?

You might be right, yeah. If we don't need to project the state it means we probably gain some performance too.

**Comment by natefaubion** [comment]
So to expand on the idea above, maybe something like this would work?
```purescript
newtype ComponentP f g s t = exists v. ComponentP
  { render :: v -> s -> Render f g v
  , eval :: Natural f (ComponentDSL f g v s t)
  , viewState :: v
  , ...
  }

type Component s f g = ComponentP f g s s

newtype Installed f g a = exists s t. Installed
  { component :: ComponentP f g s t
  , memo :: Maybe (Unit -> Tree f Unit)
  , -- projection functions from a -> s and t -> a
  }
```

So basically every component has a hidden viewState and a model state. By bundling projection functions into the installed state, you can loop back to the parent state while still remaining composable. My lens-foo is not strong enough, but I imagine it would be much like what thermite/optic-ui do. You essentially end up with separate state trees: a hidden view state tree manages completely by Halogen, and a public model tree managed by the user.

**Comment by natefaubion** [comment]
@jdegoes One thing to consider about having clear boundaries for component state, is that with the current `eval` setup, you can define async, multi-stage transitions and still mostly reason about the state. When a parent has control over child state, there's no guarantee about when the state might be swapped out from under you. This is part of the reason why in something like the Elm architecture, the transition function returns a new state and a list of effects (you have to explicitly break up a multi-stage transition).
